python/cudf/cudf/_lib/io/utils.pyx

```diff
@@ -67,3 +72,61 @@ cdef cppclass iobase_data_sink(data_sink):
 
     size_t bytes_written() with gil:
         return buf.tell()
+
+
+cdef class HostBuffer:
+    """ HostBuffer lets you spill cudf DataFrames from device memory to host
+    memory. Once in host memory the dataframe can either be re-loaded back
+    into gpu memory, or spilled to disk. This is designed to reduce the amount
+    of unnecessary host memory copies.
+
+    Examples
+    --------
+    .. code-block:: python
+      import shutil
+      import cudf
+
+      # read cudf DataFrame into buffer on host
+      df = cudf.DataFrame({'a': [1, 1, 2], 'b': [1, 2, 3]})
+      buffer = cudf.io.HostBuffer()
+      df.to_parquet(buffer)
+
+      # Copy HostBuffer back to DataFrame
+      cudf.read_parquet(df)
+
+      # Write HostBuffer to disk
+      shutil.copyfileobj(buffer, open("output.parquet", "wb"))
+    """
+    cdef vector[char] buf
+    cdef size_t pos
+
+    def __cinit__(self, int initial_capacity=0):
+        self.pos = 0
+        if initial_capacity:
+            self.buf.reserve(initial_capacity)
+
+    def __len__(self):
+        return self.buf.size()
+
+    def seek(self, size_t pos):
+        self.pos = pos
+
+    def tell(self):
+        return self.pos
+
+    def readall(self):
+        return self.read(-1)
+
+    def read(self, int n=-1):
+        if self.pos >= self.buf.size():
+            return b""
+
+        cdef size_t count = n
+        if ((n < 0) or (n > self.buf.size() - self.pos)):
+            count = self.buf.size() - self.pos
+
+        cdef size_t start = self.pos
+        self.pos += count
+
+        return PyMemoryView_FromMemory(self.buf.data() + start, count,
+                                       PyBUF_READ)
```

Comment:
Seems better to support the buffer protocol on this class instead. Otherwise this could refer to an invalid memory segment that Python has already garbage collected.

Reply:
Sounds good.

We have code already that adapts a c++ byte vector to the buffer protocol here:  https://github.com/rapidsai/cudf/blob/85ce695214759857b06690ec8d716b929b398826/python/cudf/cudf/_lib/parquet.pyx#L49-L86

Should we use that class instead of creating a new one? The only change really is between ```vector<char>``` expected by the HOST_BUFFER sink and the ```vector<uint_8>``` in this class (and I think we could either cast the cpp vector, or use cython fused types to get around).

Reply:
I think we actually have 2 instances of classes like this one where ideally we could standardize them.